### PR TITLE
domain: add recent slow queries queue and ShowSlowQuery() function

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Make user easier to debug slow queries.

### What is changed and how it works?

* Add a recent slow queries queue
* Provide `ShowSlowQuery()` function

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function `domain.ShowSlowQuery()`

Related changes

It's provided to support `admin show log` in https://github.com/pingcap/tidb/pull/7688

@winkyao @coocood

### CODE REVIEW

**Comment by shenli** [comment]
Why put it in the domain package?

**Comment by tiancaiamao** [comment]
Where should I put it ? @shenli

**Comment by tiancaiamao** [comment]
PTAL @coocood

File: domain/topn_slow_query.go [hunk 1]
```diff
@@ -59,28 +62,87 @@ func (h *slowQueryHeap) Refresh(now time.Time, recent time.Duration) {
...
+	// Queue is empty.
+	if q.tail == q.head {
+		return nil
+	}
+
+	if count > len(q.data) {
+		count = len(q.data)
+	}
+	ret := make([]*SlowQueryInfo, 0, count)
+	tail := (q.tail - 1 + len(q.data)) % len(q.data)
```
Comment (coocood) [inline domain/topn_slow_query.go]: This is simpler
```go
if q.head > q.tail {
  ret = append(ret, q.data[head:]...)
  ret = append(ret, q.data[:head]...)
} else {
  ret = append(ret, q.data[head:tail+1]...)
}
```

File: domain/topn_slow_query.go [hunk 2]
```diff
@@ -59,28 +62,87 @@ func (h *slowQueryHeap) Refresh(now time.Time, recent time.Duration) {
...
+	copy(save, h.data)
+
+	sort.Sort(h)
+
+	ret := make([]*SlowQueryInfo, 0, count)
+	for i := len(h.data) - 1; i >= 0 && len(ret) <= count; i-- {
+		ret = append(ret, h.data[i])
+	}
+
+	// Sort breaks the data, recover it here.
```
Comment (coocood) [inline domain/topn_slow_query.go]: The sorted array still maintains the heap property, we don't need to copy and recover.

Comment (tiancaiamao) [reply domain/topn_slow_query.go]: There are two problems here:
1. To keep top N slow queries, we maintains a min heap, but sort will use a reverse order
2. There is a `count` argument, so the `sorted[count]` is not the same with `heap[len]`

File: domain/topn_slow_query.go (hunk 1 above)
Comment (tiancaiamao) [reply domain/topn_slow_query.go]: It's not simpler, it brings more mental burden.

**Comment by tiancaiamao** [comment]
PTAL @coocood

File: domain/topn_slow_query.go [hunk 3]
```diff
@@ -40,11 +43,11 @@ func (h *slowQueryHeap) Pop() interface{} {
 	return x
 }
 
-func (h *slowQueryHeap) Refresh(now time.Time, recent time.Duration) {
+func (h *slowQueryHeap) Refresh(now time.Time, period time.Duration) {
```
Comment (coocood) [inline domain/topn_slow_query.go]: Why it is exported?

Comment (tiancaiamao) [reply domain/topn_slow_query.go]: This is not exported, `slowQueryHeap` is in lower case

**Comment by tiancaiamao** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @winkyao

File: domain/topn_slow_query.go [hunk 4]
```diff
@@ -101,9 +155,47 @@ func (q *topNSlowQueries) Append(info *SlowQueryInfo) {
...
+	case "internal":
+		ret = q.internal.Query(count)
+	case "all":
+		tmp := make([]*SlowQueryInfo, 0, len(q.user.data)+len(q.internal.data))
+		tmp = append(tmp, q.user.data...)
+		tmp = append(tmp, q.internal.data...)
+		tmp1 := slowQueryHeap{tmp}
+		sort.Sort(&tmp1)
+		ret = make([]*SlowQueryInfo, 0, count)
+		for i := len(tmp) - 1; i >= 0 && len(ret) <= count; i-- {
```
Comment (crazycs520) [inline domain/topn_slow_query.go]: len(ret) < count ?

File: ast/misc.go [hunk 5]
```diff
@@ -626,6 +626,26 @@ type HandleRange struct {
 	End   int64
 }
 
+// ShowLogType defines the type for SlowLog.
+type ShowLogType int
```
Comment (zz-jason) [inline ast/misc.go]: s/ShowLogType/ShowSlowQueryType/

File: ast/misc.go [hunk 6]
```diff
@@ -626,6 +626,26 @@ type HandleRange struct {
...
+	// ShowLogTop is a ShowLogType constant.
+	ShowLogTop ShowLogType = iota
+	// ShowLogRecent is a ShowLogType constant.
+	ShowLogRecent
+)
+
+// ShowLog is used for the following command:
+//	admin show log top [user | internal | all] N
+//	admin show log recent N
+type ShowLog struct {
```
Comment (zz-jason) [inline ast/misc.go]: s/ShowLog/ShowQueries/

File: ast/misc.go (hunk 5 above)
Comment (tiancaiamao) [reply ast/misc.go]: When I merge master, this would be solved.

File: ast/misc.go (hunk 6 above)
Comment (tiancaiamao) [reply ast/misc.go]: ditto

**Comment by tiancaiamao** [comment]
PTAL @crazycs520 @zz-jason

File: domain/topn_slow_query.go [hunk 7]
```diff
@@ -59,28 +62,79 @@ func (h *slowQueryHeap) Refresh(now time.Time, recent time.Duration) {
...
+	size int
+}
+
+func (q *slowQueryQueue) Enqueue(info *SlowQueryInfo) {
+	if len(q.data) < q.size {
+		q.data = append(q.data, info)
+		return
+	}
+
+	q.data = append(q.data, info)[1:]
```
Comment (winkyao) [inline domain/topn_slow_query.go]: If you just use slice here, the q.data will keep the continuous memory, it may cause oom. What about using a list?

Comment (winkyao) [reply domain/topn_slow_query.go]: Do you do a large amount insert test?

Comment (tiancaiamao) [reply domain/topn_slow_query.go]: @coocood suggest me to use this.
The code is the simplest queue implementation, although it double the memory usage.

Comment (tiancaiamao) [reply domain/topn_slow_query.go]: The queue size will be 500.
We just keep the most recent queries in the queue.

**Pull request merged**